### PR TITLE
David/multiple proxy test

### PR DESCRIPTION
**Pull request by WyseNynja**
- Should fix influx vs mysql discrepancies and balance bugs. 
- Must still test_users

### CODE REVIEW

File: web3_proxy/tests/test_multiple_proxy.rs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    admin_increase_balance(&x_1, &r, &admin_login, &user_1_wallet, Decimal::from(30)).await;
+    // Just check that balances were properly distributed
+    let user_0_balance_response = user_get_balance(&x_0, &r, &user_0_login).await;
+    let user_0_balance_pre = user_0_balance_response.remaining();
+    let user_1_balance_response = user_get_balance(&x_1, &r, &user_1_login).await;
+    let user_1_balance_pre = user_1_balance_response.remaining();
+
+    // Make sure they both have balance now
+    assert_eq!(user_0_balance_pre, Decimal::from(20));
+    assert_eq!(user_1_balance_pre, Decimal::from(20));
```
Comment (WyseNynja) [inline web3_proxy/tests/test_multiple_proxy.rs:64]: This will be $30

File: web3_proxy/tests/test_multiple_proxy.rs [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    let user_1_secret_key = user_get_first_rpc_key(&x_1, &r, &user_1_login)
+        .await
+        .secret_key;
+
+    for _ in 1..number_requests {
+        let proxy_0_url = x_0.proxy_provider.url().clone();
+        handles.push(tokio::spawn(async move {
+            // Because of the move, we re-create the provider each time; this should be fine because it's http
+            let proxy_provider_0 =
+                create_provider_for_user(proxy_0_url.clone(), user_0_secret_key.clone()).await;
```
Comment (WyseNynja) [inline web3_proxy/tests/test_multiple_proxy.rs:88]: Do this outside the loop and put it in an Arc. Then you can clone it

File: web3_proxy/tests/test_multiple_proxy.rs [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                .await
+                .unwrap()
+                .unwrap()
+        }));
+    }
+
+    // Flush all stats here
+    let flush_count = x_0.flush_stats().await.unwrap();
+    assert_eq!(flush_count.timeseries, 0);
+    assert!(flush_count.relational > 0);
```
Comment (WyseNynja) [inline web3_proxy/tests/test_multiple_proxy.rs:110]: assert the actual number expected

File: web3_proxy/tests/test_multiple_proxy.rs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+    let flush_count = x_1.flush_stats().await.unwrap();
+    assert_eq!(flush_count.timeseries, 0);
+    assert!(flush_count.relational > 0);
```
Comment (WyseNynja) [inline web3_proxy/tests/test_multiple_proxy.rs:113]: Validate the actual number expected

**Comment by WyseNynja** [comment]
The test config has $.10/cu costs, so you might want more than $20 or $30

File: web3_proxy/tests/test_multiple_proxy.rs [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    assert_eq!(
+        mysql_stats["error_response"],
+        influx_stats["error_response"]
+    );
+    assert_eq!(
+        mysql_stats["archive_needed"],
+        influx_stats["archive_needed"]
+    );
+    assert_eq!(
+        Decimal::from_str(&mysql_stats["chain_id"].to_string().replace('"', "")).unwrap(),
```
Comment (WyseNynja) [inline web3_proxy/tests/test_multiple_proxy.rs:166]: This should be a u64, not a decimal. Though I guess it works either way

File: web3_proxy/tests/test_multiple_proxy.rs [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+                .replace('"', "")
+        )
+        .unwrap()
+    );
+    assert_eq!(
+        Decimal::from_str(&mysql_stats["backend_requests"].to_string()).unwrap(),
+        Decimal::from_str(&influx_stats["total_backend_requests"].to_string()).unwrap()
+    );
+
+    // balance with balance
```
Comment (WyseNynja) [inline web3_proxy/tests/test_multiple_proxy.rs:235]: What does "balance with balance" mean?

**Review by WyseNynja — COMMENTED** [review]
So glad this reproduced the issue

File: web3_proxy/src/stats/influxdb_queries.rs [hunk 7]
```diff
@@ -88,8 +88,10 @@ pub async fn query_user_stats<'a>(
         ));
     }
 
+    info!("Getting influx client");
```
[the remark this answers is not in the document]
Comment (yenicelik) [reply web3_proxy/src/stats/influxdb_queries.rs]: I forgot to remove these, we dont need these

File: web3_proxy/tests/common/influx.rs [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+            .expect("failed to start influx");
+
+        // original port 18086
+        info!("Creation command is: {:?}", cmd);
+
+        // give the db a second to start
+        // TODO: wait until docker says it is healthy
+        sleep(Duration::from_secs(1)).await;
+
+        // TODO: why is this always empty?!
```
[the remark this answers is not in the document]
Comment (yenicelik) [reply web3_proxy/tests/common/influx.rs:91]: Removed these from mysql and influx.rs

File: web3_proxy/src/stats/mod.rs [hunk 9]
```diff
@@ -488,6 +491,13 @@ impl BufferedRpcQueryStats {
             builder = builder.tag("rpc_secret_key_id", rpc_secret_key_id.to_string());
         }
```
[the remark this answers is not in the document]
Comment (yenicelik) [reply web3_proxy/src/stats/mod.rs:493]: Ok I'll try again without static variables

**Pull request merged**